packages/repeater/src/bus/DefaultRepeaterServer.ts

```diff
@@ -0,0 +1,311 @@
+  readonly proxyUrl?: string;
+  readonly insecure?: boolean;
+}
+
+export const DefaultRepeaterServerOptions: unique symbol = Symbol(
+  'DefaultRepeaterServerOptions'
+);
+
+export const enum SocketEvents {
+  DEPLOYED = 'deployed',
+  DEPLOY = 'deploy',
+  UNDEPLOY = 'undeploy',
+  UNDEPLOYED = 'undeployed',
+  ERROR = 'error',
+  UPDATE_AVAILABLE = 'update-available',
+  PING = 'ping',
+  REQUEST = 'request'
+}
+
+interface SocketListeningEventMap {
+  [SocketEvents.DEPLOYED]: (event: RepeaterServerDeployedEvent) => void;
+  [SocketEvents.UNDEPLOYED]: () => void;
+  [SocketEvents.ERROR]: (event: RepeaterServerErrorEvent) => void;
+  [SocketEvents.UPDATE_AVAILABLE]: (
+    event: RepeaterUpgradeAvailableEvent
+  ) => void;
+  [SocketEvents.REQUEST]: (
+    request: RepeaterServerRequestEvent,
+    callback: CallbackFunction<RepeaterServerRequestResponse>
+  ) => void;
+}
+
+interface SocketEmitEventMap {
+  [SocketEvents.DEPLOY]: (options: DeployCommandOptions) => void;
+  [SocketEvents.UNDEPLOY]: () => void;
+  [SocketEvents.PING]: () => void;
+}
+
+@scoped(Lifecycle.ContainerScoped)
+@injectable()
+export class DefaultRepeaterServer implements RepeaterServer {
+  private readonly MAX_DEPLOYMENT_TIMEOUT = 60_000;
+  private readonly MAX_PING_INTERVAL = 10_000;
+  private readonly MAX_RECONNECTION_ATTEMPTS = 20;
+  private readonly MIN_RECONNECTION_DELAY = 1000;
+  private readonly MAX_RECONNECTION_DELAY = 86_400_000;
+
+  private latestReconnectionError?: Error;
+  private pingTimer?: Timer;
+  private connectionTimer?: Timer;
+  private _socket?: Socket<SocketListeningEventMap, SocketEmitEventMap>;
+  private connectionAttempts = 0;
+
+  private get socket() {
+    if (!this._socket) {
+      throw new Error(
+        'Please make sure that repeater established a connection with host.'
+      );
+    }
+
+    return this._socket;
+  }
+
+  constructor(
+    private readonly logger: Logger,
+    @inject(RepeaterEventHub)
+    public readonly events: RepeaterEventHub,
+    @inject(DefaultRepeaterServerOptions)
+    private readonly options: DefaultRepeaterServerOptions
+  ) {}
+
+  public disconnect() {
+    this.events.removeAllListeners();
+    this.clearPingTimer();
+    this.clearConnectionTimer();
+
+    this._socket?.disconnect();
+    this._socket?.removeAllListeners();
+    this._socket = undefined;
+  }
+
+  public async deploy(
+    options: DeployCommandOptions
+  ): Promise<RepeaterServerDeployedEvent> {
+    process.nextTick(() => this.socket.emit(SocketEvents.DEPLOY, options));
+
+    const [result]: RepeaterServerDeployedEvent[] = await Promise.race([
+      once(this.socket, SocketEvents.DEPLOYED),
+      new Promise<never>((_, reject) =>
+        setTimeout(
+          reject,
+          this.MAX_DEPLOYMENT_TIMEOUT,
+          new Error('No response.')
+        ).unref()
+      )
+    ]);
+
+    this.createPingTimer();
+
+    return result;
+  }
+
+  public async connect(repeaterId: string) {
```

Comment:
We don't have a repeater ID at this moment; repeater will be created only on `deploy`.

```suggestion
  public async connect(domain: string = hostname()) {
```